src/csmake.pl

```diff
@@ -611,6 +683,62 @@ sub abs_if_exists
 exit system(($real, @ARGV)) / 256;
 #@END
 
+#@BEGIN spy-install
+#
+# Spy on install invocations and construct corresponding CScout directives
+#
+
+$real = which($0);
+
+$origline = "install " . join(' ', @ARGV);
+$origline =~ s/\n/ /g;
+
+@ARGV2 = @ARGV;
+
+my @excecutables = ();
+my $dest;
+my $length = @ARGV2;
+my $counter = 0;
+
+while (my $i = shift @ARGV2) {
+	$counter++;
+	# If -d option is used skip this command
+	if ($i eq "-d") {
+		last;
+	}
+	# If -t save next argument to directory
+	if ($i eq "-t") {
+		$dest = shift @ARGV2;
+		$counter++;
+	}
+	if ($counter == $length) {
+		if (length $dest) {
+			if (-x $i && ! -d $i) {
+				push(@excecutables, $i);
+			}
+		} else {
+			$dest = $i;
+		}
+		last;
+	}
+	# If file is a real executable add it to array
+	if (-x $i && ! -d $i) {
+		push(@excecutables, $i);
+	}
+}
+
+if (@excecutables) {
+	close RULES;
+	foreach (@excecutables) {
+		prepend_rules("INSTALL $_ $dest\n");
```

Comment:
Given that you are post-processing the rules, why to do you the expensive O(N*N) prepend operation?

Reply:
We want to read the install rules first in the processing phase, I suppose we could achieve the same behaviour if we parse the rules two times. In the first pass we'll only read the INSTALL rules. Should I change it?

Reply:
You can store them in the first pass in a suitable in-memory data structure.

Reply:
By doing that we will read the INSTALL rules at the end of the processing phase, whereas we want to read them before processing the projects.

Reply:
OK, two passes then. (Use rewind).

Reply:
Done